cs3342_09-master/README.md

```diff
@@ -0,0 +1,191 @@
+.pdf) and put its file name as the answer.
+
+For question 2.5, you'll write one or more source code files and add them to the
+repository. You will then show the command used to run the files in the answer.
+Do _not_ upload any binary files for this section.
+
+Please do not edit anything outside the answers sections.
+
+
+# Q1: Mr Fussy Builds a Path
+
+# Q1
+
+Mr Fussy wants to build a path using a row of large square pavers. His path is
+one paver wide and an odd number of pavers long. Pavers come in three colors:
+red, green, blue.
+
+Mr Fussy has a rule: the sequence of colors must be symmetrical across the
+length of the path: the last paver must be the same color as the first, the
+second to last the same color as the second, and so on.
+
+This is a valid path:  rgbgr
+So is this: rrr
+
+This is not: rgb
+This is not: rggr    (it must be an odd length)
+
+## Q1.1  (5 points)
+
+Write a Chomsky type 2 grammar that describes any valid path containing at
+least one paver. Use `S` as the start state, and `r`, `g`, `b` as terminals that
+represent the tiles.
+
+## A1.1
+
+
+S -> color1 | color2 S color2
+
+color -> r|g|b
+
+## Q1.2  (1 point for the O() answer, 2 for the sentence)
+
+Using big-O notation, what is the likely memory requirement for a parser that
+can validate a particular path configuration, where `n` is the number of tiles?
+In one sentence, explain why.
+
+## A1.2
+
+O(1) becuase the parser would only need to iterate over the path one time 
+
+
+# Q2
+
+A simple sentence structure might be
+
+    "the" (zero or more adjectives) noun verb (optional adverb)
+
+Adjectives are "lazy" or "smelly"
+
+Nouns are "dog" or "cat"
+
+Verbs are "ate" and "ran"
+
+Adverbs are "slowly" and "noisily"
+
+The following are examples of valid sentences:
+
+* The smelly dog ran.
+* The smelly dog ran slowly.
+* The cat ate noisily.
+
+## Q2.1 (7 points)
+
+Write the BNF (not EBNF) description for this language.
+
+## A2.1
+
+\<sentence> ::= "the" + \<adj> + \<noun> + \<verb> + \<adverb>
+\<adj> ::= \<<empty>> | "lazy" | "smelly"
+\<adverb> ::=  \<<empty>> | "slowly" | "noisily"
+\<noun> ::=   "dog" | "cat"
+\<verb> ::=   "ate" | "ran"
+
+
+## Q2.2 (5 points)
+
+Write this grammar using EBNF with common extensions
+
+## A2.2
+
+\<sentence> ::= "the" + {\<adj>} + \<noun> + \<verb> + [\<adverb>]
+\<adj> ::= "lazy" | "smelly
+\<adverb> ::= "slowly" | "noisily"
+\<noun> ::=   "dog" | "cat"
+\<verb> ::=   "ate" | "ran"
+
+
+## Q2.3 (6 points)
+
+  Draw a diagram for an FSM which recognizes these sentences. Use EOI as the
+  event that occurs at the end-of-input. Start with a state named S0. Receiving
+  the word "the" in that state will transition to state S1. Name the final state
+  END.
+
+  You can hand draw it and snap a picture, or use a drawing tool. Either way,
+  upload the image, and put the file name in the answer below.
+
+  (Hint: my answer has seven states including the start and end states)
+
+
+## A2.3
+
+Name = A2.3
```

Comment:
I think the labels on some of the transitions are incorrect: S1 -> S2 should be "noun", and S2 -> S3 should be verb. Then S3 should go to the END state on EOI, and to S4 on seeing an adverb. S4 should also go to the end state on EOI

![image](https://user-images.githubusercontent.com/10648/78714182-0bd98180-78e1-11ea-9e21-3bb6a1bcc01b.png)


# -3